python/ray/tune/tests/test_sample.py

```diff
@@ -1023,6 +1023,23 @@ def testConvertOptuna(self):
             param.suggest_loguniform("b/z", 1e-4, 1e-2)
         ]
 
+        def optuna_define_by_run(ot_trial):
+            ot_trial.suggest_categorical("a", [2, 3, 4])
+            ot_trial.suggest_int("b/x", 0, 5, 2)
+            ot_trial.suggest_loguniform("b/z", 1e-4, 1e-2)
```

Comment:
I'm wondering if we could use a hierarchical search space somewhere here

Reply:
How do you think it could look like?

Reply:
Maybe something like this?
```
a = ot_trial.suggest_categorical("a", [2, 3, 4])
if a < 3:
   b = ot_trial.suggest_int("a/b", 0, 3)
   c = ot_trial.suggest_int("a/first", 2, 8)
else:
     b = ot_trial.suggest_int("a/b", 4, 10)
     c = ot_trial.suggest_uniform("a/second", -0.4, 0.4)
```

The we would check two cases: Same parameter name but different bounds, and different parameter names for different branches

Reply:
This example will actually raise an error as such a dict can't be unflattened (it will try to assign child dicts to "a", but it's already an int). I've modified `unflatten_dict` util func to raise an informative exception and added tests for it.